cou/apps/app.py

```diff
@@ -245,24 +224,121 @@ def __str__(self) -> str:
+        """Populate application units."""
+        if not self.is_subordinate:
+            for name, unit in self.status.units.items():
+                compatible_os_version = (
+                    self._get_latest_compatible_os_version_by_unit_workload_version(unit)
+                )
+                self.units.append(
+                    ApplicationUnit(
+                        name=name,
+                        workload_version=unit.workload_version,
+                        os_version=compatible_os_version,
+                        machine=unit.machine,
+                    )
+                )
+
+    def _get_latest_compatible_os_version_by_unit_workload_version(
+        self, unit: UnitStatus
+    ) -> OpenStackRelease:
+        """Get the latest compatible OpenStack release based on the unit workload version.
+
+        If the application is versionless, then get from the current OpenStack release.
+
+        :param unit: Application Unit
+        :type unit: UnitStatus
+        :raises ApplicationError: When there are no compatible OpenStack release for the
+        workload version.
+        :return: The latest compatible OpenStack release.
+        :rtype: OpenStackRelease
+        """
+        if self.is_os_channel_based:
+            return self.channel_codename
+
+        try:
+            return max(
+                OpenStackCodenameLookup.find_compatible_versions(self.charm, unit.workload_version)
+            )
+        except ValueError as exc:
+            raise ApplicationError(
+                f"'{self.name}' with workload version {unit.workload_version} has no "
+                f"compatible OpenStack release."
+            ) from exc
+
+    @property
+    def channel(self) -> str:
+        """Get charm channel of the application.
+
+        :return: Charm channel. E.g: ussuri/stable
+        :rtype: str
+        """
+        return self._channel
+
+    @channel.setter
+    def channel(self, charm_channel: str) -> None:
+        """Set charm channel of the application.
+
+        When application comes from charm store, the channel won't be OpenStack related.
+        :param charm_channel: Charm channel. E.g: ussuri/stable
+        :type charm_channel: str
+        :raises ValueError:  Exception raised when channel is not a valid OpenStack
+            channel.
+        """
+        if (
+            self.is_from_charm_store
+            or (not self.is_os_channel_based and self.is_subordinate)
+            or self.is_valid_track
+        ):
+            self._channel = charm_channel
+            return
+        raise ValueError(self.channel_err_msg(charm_channel))
+
+    def channel_err_msg(self, charm_channel: str) -> str:
+        """Error message when channel is not valid.
+
+        :param charm_channel: Charm channel.
+        :type charm_channel: str
+        :return: error message
+        :rtype: str
+        """
+        return f"'{self.name}' has invalid channel: '{charm_channel}'"
```

Comment:
To me this is unnecessary function used once, so instead of one line we add 10+unit test.

Reply:
All the logic of the `@channel.setter` and channel property is the same for `OpenStackApplication` and `OpenStackAuxiliaryApplication` with the exception of this method. 
I just want to have a more clear error msg for the `AuxiliaryApplication` that depends on the `charm`, `series` and `os_release`.

In the end of the day I think it's a good deal to share the same logic on both classes and having this method to be more explicit in the error message.

Reply:
To me this with all functions like this, it's just harder to understand what is happening in one class. It's harder to debug if something is not working, because logic is across multiple classes. It's good practice until it become unreadable.
I would rather change it in two places and be explicit, than jump from one class to another to get idea what is actually happening and especially when function is one line of code.